zilliqa/src/uccb/client.rs

```diff
@@ -0,0 +1,62 @@
+use std::sync::Arc;
+
+use anyhow::Result;
+use ethers::{
+    middleware::{MiddlewareBuilder, NonceManagerMiddleware, SignerMiddleware},
+    providers::{Middleware, Provider, Ws},
+    signers::{LocalWallet, Signer},
+    types::{Address, U256},
+};
+
+use crate::uccb::cfg::ChainConfig;
+
+pub type Client = NonceManagerMiddleware<SignerMiddleware<Provider<Ws>, LocalWallet>>;
+
+#[derive(Debug, Clone)]
+pub struct ChainClient {
+    pub rpc_url: String,
+    pub client: Arc<Client>,
+    pub validator_manager_address: Address,
+    // pub chain_gateway_address: Address,
+    pub chain_id: U256,
+    pub wallet: LocalWallet,
+    pub chain_gateway_block_deployed: u64,
+    pub block_instant_finality: bool,
+    pub legacy_gas_estimation: bool,
+}
+
+impl ChainClient {
+    pub async fn new(
+        config: &ChainConfig,
+        validator_manager_address: Address,
+        // chain_gateway_address: Address,
+        wallet: LocalWallet,
+    ) -> Result<Self> {
+        // let provider = Provider::<Http>::try_from(config.rpc_url.as_str())?;
+        let ws = Ws::connect(&config.rpc_url).await?;
+        let provider = ethers::providers::Provider::<Ws>::new(ws);
+        let chain_id = provider.get_chainid().await?;
+
+        let client: Arc<Client> = Arc::new(
+            provider
+                .with_signer(wallet.clone().with_chain_id(chain_id.as_u64()))
+                .nonce_manager(wallet.address()),
+        );
+
+        // TODO: get the validator_manager_address from chain_gateway itself
+        // let chain_gateway = ChainGateway::new(config.chain_gateway_address, client.clone());
+        // let validator_manager_address: Address = chain_gateway.validator_manager().call().await?;
+
+        Ok(ChainClient {
+            rpc_url: config.rpc_url.clone(),
+            client,
+            validator_manager_address,
+            // chain_gateway_address,
+            chain_id,
+            wallet,
+            chain_gateway_block_deployed: config.chain_gateway_block_deployed,
+            block_instant_finality: config.block_instant_finality.unwrap_or_default(),
+            legacy_gas_estimation: config.legacy_gas_estimation.unwrap_or_default(),
```

Comment:
It might be nicer to assign these defaults in the config itself (when deserialising via `serde(default = ...)`)?